Photo Mode

TrainWrack opened the pull request:
## Checklist

- [x] I have added a changelog entry to CHANGELOG.md file on the branch/fork (if it is an internal change then it is not needed) 
- [x] Pull request meets the Coding Conventions standards: https://github.com/MontyTRC89/TombEngine/blob/master/CONTRIBUTING.md#coding-conventions

## Links to issue(s) this pull request concerns (if applicable)



## Description of pull request 

Adds photo mode

```
require("Engine.PhotoMode.PhotoMode")
```
Controls
<img width="494" height="317" alt="image" src="https://github.com/user-attachments/assets/5b6dd88c-8dee-442d-bd5b-038b0c434330" />

[PhotoModeSprites.zip](https://github.com/user-attachments/files/29138183/PhotoModeSprites.zip)






Use provided wad. Once the PR is ready for merging I will attach another wad and remove entries from outfits and expressions.

TrainWrack committed:
WIP

TrainWrack committed:
Remove unnecessary code

TrainWrack committed:
WIP

TrainWrack committed:
Limit roll to object camera

TrainWrack committed:
Finish

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Changelog update

Lwmte commented on TombEngine/Game/camera.cpp:
```diff
@@ -212,6 +212,16 @@ short GetCurrentFOV()
 	return CurrentFOV;
 }
 
+void AlterRoll(short value)
```
Not sure if we need two layers of one-line functions. Can you directly modify/get roll within script module?

TrainWrack replied:
Its ok. Its better than accessing Camera directly in script functions and follows convention for FoV

Lwmte commented on TombEngine/Scripting/Internal/ReservedScriptNames.h:
```diff
@@ -339,6 +339,8 @@ static constexpr char ScriptReserved_FadeOutComplete[]				= "FadeOutComplete";
 static constexpr char ScriptReserved_SetCineBars[]					= "SetCineBars";
 static constexpr char ScriptReserved_SetFOV[]						= "SetFOV";
 static constexpr char ScriptReserved_GetFOV[]						= "GetFOV";
+static constexpr char ScriptReserved_SetCameraRoll[]				= "SetObjectCameraRoll";
```
Does roll setting/getting only works in object camera mode? Can we make it work in other modes as well? Otherwise this API looks a little inconsistent (which underlines general problem with lack of game camera API though).

TrainWrack replied:
Sorted

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Finish Roll

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Fix docs

Lwmte committed:
Fixed #1952

Lwmte committed:
Fixed empty inventory screen after exiting examine mode

TrainWrack committed:
Fix GetCurrentRoll

TrainWrack committed:
Fix WALL_SWITCH and CEILING_SWITCH enums

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Start Photo Mode

TrainWrack committed:
Strings Cleanup

TrainWrack committed:
Cleanup

TrainWrack committed:
FIx camera

TrainWrack committed:
Layout changes

TrainWrack committed:
Reset Hair
Get/Set Holsters
Fix Object Camera

TrainWrack committed:
WIP

TrainWrack committed:
Allow lights in Freeze Mode

TrainWrack committed:
WIP

TrainWrack committed:
Add gunflashes

Feature complete

TrainWrack committed:
WIP

TrainWrack committed:
WIP

TrainWrack committed:
Merge remote-tracking branch 'upstream/lwmte/bugfixes_v20' into ExposeCameraRoll

TrainWrack committed:
WIP

TrainWrack committed:
Fix Gunflash

TrainWrack committed:
GetMeshSwap ObjectID
Cleanup

TrainWrack committed:
Finish Apply Outfit

TrainWrack committed:
Fix function

TrainWrack committed:
WIP

TrainWrack committed:
Finish skin functions

TrainWrack committed:
Update the camera code
Update outfit modes

TrainWrack committed:
Outfit mode rewrite

TrainWrack committed:
Fix weapons and expressions

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
WIP

TrainWrack commented on Scripts/Engine/PhotoMode/States.lua:
```diff
@@ -0,0 +1,296 @@
+    end)
+
+    TEN.View.SetFOV(snap.fov)
+    TEN.View.SetRoll(0)
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/Settings.lua:
```diff
@@ -0,0 +1,343 @@
+    scale     = TEN.Vec2(100, 100),
+    alignMode = TEN.View.AlignMode.CENTER,
+    scaleMode = TEN.View.ScaleMode.STRETCH,
+    blendMode = TEN.Effects.BlendID.ALPHA_BLEND,
```
Its the updated Enum

TrainWrack commented on Scripts/Engine/PhotoMode/Settings.lua:
```diff
@@ -0,0 +1,343 @@
+
+local RingSettings = require("Engine.RingInventory.Settings")
+
+local Settings = {}
```
Ring Inventory sits outside the scope of this module.

TrainWrack commented on Scripts/Engine/PhotoMode/States.lua:
```diff
@@ -0,0 +1,296 @@
+    snap.laraState    = Lara:GetState()
+
+    snap.fov  = TEN.View.GetFOV()
+    snap.roll = 0
```
Fixed

TrainWrack commented on TombEngine/Scripting/Internal/TEN/Objects/Moveable/MoveableObject.cpp:
```diff
@@ -1139,6 +1180,14 @@ void Moveable::UnswapSkinnedMesh()
+// @function Moveable:ClearSkinnedMesh
+void Moveable::ClearSkinnedMesh()
+{
+	int realID = _moveable->ObjectNumber == GAME_OBJECT_ID::ID_LARA ? GAME_OBJECT_ID::ID_LARA_SKIN : _moveable->ObjectNumber;
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,983 @@
+    local walkHeld = TEN.Input.IsKeyHeld(TEN.Input.ActionID.Q)
+    local invHeld  = TEN.Input.IsKeyHeld(TEN.Input.ActionID.INVENTORY)
+
+    if walkHeld or invHeld then
```
To be decided how to open Photo Mode

TrainWrack commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -636,6 +638,75 @@ std::unique_ptr<Moveable> LaraObject::GetPlayerInteractedMoveable() const
+// @tparam Objects.WeaponType weaponType Weapon type to spawn the flash for.
+// @usage
+// Lara:SpawnGunFlash(WeaponType.PISTOLS)
+void LaraObject::SpawnGunFlash(LaraWeaponType weaponType)
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,983 @@
+    if preset.weaponType == TEN.Objects.WeaponType.FLARE then
+        local flare = Flow.GetSettings().Flare
+        local position = Lara:GetJointPosition(preset.meshIndices[1], flare.offset)
+        pcall(function() Effects.EmitLight(position, flare.color, flare.range) end)
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/Input.lua:
```diff
@@ -0,0 +1,196 @@
+    end
+
+    Lara:SetPosition(newPos)
+    pcall(function() Lara:ResetHair() end)
```
Its intentional. Added a check for reset hair to only happen if position or rotation changes.

TrainWrack commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,772 @@
+            menu.titleString, pos, menu.titleTextScale,
+            ColorCombine(menu.titleTextColor, actualTransparency),
+            menu.titleTranslate, menu.titleTextFlags
+        )
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,772 @@
+                menu.itemsTextScale,
+                ColorCombine(menu.itemsTextColor, actualTransparency),
+                menu.itemsTranslate
+            )
```
FIxed

TrainWrack commented on Scripts/Engine/PhotoMode/Settings.lua:
```diff
@@ -0,0 +1,343 @@
+{
+    plainText = Flow.GetSettings().UI.plainTextColor,
+    headerText = Flow.GetSettings().UI.headerTextColor,
+    optionText = Flow.GetSettings().UI.optionTextColor,
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,772 @@
+                menu.optionsTextScale,
+                ColorCombine(menu.optionsTextColor, actualTransparency),
+                menu.optionsTranslate
+            )
```
Fixed

TrainWrack commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,772 @@
+-- @module Engine.PhotoMode.Menu
+-- @local
+
+local InputHelpers = require("Engine.RingInventory.InputHelpers")
```
Ring Inventory is now merged to develop

TrainWrack commented on TombEngine/Game/camera.cpp:
```diff
@@ -234,6 +234,16 @@ short GetCurrentFOV()
+
+float GetCurrentRoll()
+{
+	return FROM_RAD(Camera.Roll);
```
Fixed

TrainWrack commented on TombEngine/Game/camera.h:
```diff
@@ -95,6 +95,8 @@ void LookCamera(ItemInfo& item, const CollisionInfo& coll);
 void AlterFOV(short value, bool store = true);
 short GetCurrentFOV();
+void AlterRoll(short value);
+float GetCurrentRoll();
```
fixed

TrainWrack committed:
Address bugs

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Add GetPostProcess

TrainWrack committed:
Adjust indent

TrainWrack committed:
Add settings reset
Post Process, DOF reset

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Adjust as per GetPostProcess

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
WIP

TrainWrack committed:
WIP

TrainWrack committed:
Finish Photo Mode

TrainWrack committed:
Fix menu acceleration

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Add slots

TrainWrack committed:
Add intensity and combine options

TrainWrack committed:
Add color wheel

TrainWrack committed:
Cleanup

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

Copilot committed:
Replace SkinIndex with SkinObjectID/SkinSwapIndex in MoveableModelData

Agent-Logs-Url: https://github.com/TrainWrack/TombEngine/sessions/f5ff9f2c-1fd9-4c12-9ab1-4409fef76b81

Co-authored-by: TrainWrack <120750885+TrainWrack@users.noreply.github.com>

Copilot committed:
Fix alignment in MoveableModelData and simplify tuple return

Agent-Logs-Url: https://github.com/TrainWrack/TombEngine/sessions/f5ff9f2c-1fd9-4c12-9ab1-4409fef76b81

Co-authored-by: TrainWrack <120750885+TrainWrack@users.noreply.github.com>

TrainWrack committed:
Update branch

TrainWrack committed:
Merge pull request #17 from TrainWrack/copilot/update-skin-index-management

Refactor: Store skin swap as (SkinObjectID, SkinSwapIndex) instead of raw global mesh index

TrainWrack committed:
Update

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
WIP

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Finish accessory mode

TrainWrack committed:
WIP Controls and Docs

TrainWrack committed:
WIP

TrainWrack committed:
finish controls
Need to finish the camera jutter
adjust mouse sensitivity

TrainWrack committed:
Fix camera pitch

TrainWrack committed:
WIP Comments

TrainWrack committed:
WIP Lua docs

TrainWrack committed:
WIP

TrainWrack committed:
finish photo mode

TrainWrack committed:
Merge branch 'develop' into ExposeCameraRoll

TrainWrack committed:
Adjust movement speed to account for auto interpolation disabling

Lwmte commented on TombEngine/Game/items.h:
```diff
@@ -105,13 +105,16 @@ struct MoveableAnimData
-	int BaseMesh  = 0;
-	int SkinIndex = 0;
+	int BaseMesh      = 0;
+	int SkinObjectID  = NO_VALUE;
```
In my opinion, it's over-engineering. I don't see why builder may want to pack several skins into a single slot. In worst case, they can place different skins into different slots, but abusing unskinned meshes for keeping different outfits feel too much for just a photo mode.

TrainWrack replied:
As  discussed this is needed to store both the values if user uses

https://tombengine.github.io/3%20classes/Objects.Moveable.html#Moveable:SwapSkinnedMesh 

to set a skin from a mesh index

Lwmte commented on TombEngine/Objects/Generic/generic_objects.cpp:
```diff
@@ -202,7 +202,7 @@ void StartSwitches(ObjectInfo* object)
 		object->shadowType = ShadowMode::All;
 	}
 
-	for (int objectID = ID_UNDERWATER_WALL_SWITCH1; objectID <= ID_UNDERWATER_WALL_SWITCH2; objectID++)
```
I thought this name change was already done via another PR?

TrainWrack replied:
Yeah I merged the bugfix PR. It should be ok though once that PR gets merged with develop.

Lwmte commented on TombEngine/Objects/game_object_ids.h:
```diff
@@ -42,8 +42,9 @@ enum GAME_OBJECT_ID : short
-	
-	ID_SNOWMOBILE_TRACKS = 49,
+
+	ID_PHOTOMODE_ANIMS = 48,
```
Why slot ID#48 was selected for keeping photo mode animations? To me, it's more logical to keep it in slot 39.

TrainWrack replied:
Addressed

Lwmte commented on TombEngine/Game/Lara/lara_fire.cpp:
```diff
@@ -440,6 +441,92 @@ HolsterSlot GetWeaponHolsterSlot(LaraWeaponType weaponType)
 	}
 }
 
+LaraWeaponType GetHolsterSlotWeapon(HolsterSlot slot)
```
This function is a script-only function. Therefore, I suggest to isolate it to LaraObject.cpp.

TrainWrack replied:
Addressed

Lwmte commented on TombEngine/Game/Lara/lara_fire.cpp:
```diff
@@ -440,6 +441,92 @@ HolsterSlot GetWeaponHolsterSlot(LaraWeaponType weaponType)
+	}
+}
+
+void SpawnWeaponFlash(ItemInfo& laraItem, LaraWeaponType weaponType)
```
As above. Also, not sure if this function is similar to old flipeffects that serve the same purpose? Flipeffect names are `ShootRightGun` and `ShootLeftGun`

TrainWrack replied:
Flipeffects won't work as they work with WeaponType

Lwmte commented on TombEngine/Game/Lara/lara_struct.h:
```diff
@@ -1152,9 +1152,10 @@ struct ArmInfo
+	int				GunFlash	 = 0;
+	int				GunSmoke	 = 0;
+	int				AimDelay	 = 0;
+	LaraWeaponType	GunFlashType = LaraWeaponType::None;
```
Again, I believe this is over-engineering. We introduce ambiguity here by introducing `GunFlashType` in addition to `Lara.Control.Weapon.GunType`. I don't understand in which circumstances first one is used, and in which circumstances another one is used.

TrainWrack replied:
Have used GunFlashType across the codebase for transparency and consistency

Lwmte commented on TombEngine/Game/gui.cpp:
```diff
@@ -3502,7 +3502,7 @@ namespace TEN::Gui
 		if (GuiIsDeselected())
 		{
 			SoundEffect(SFX_TR4_MENU_SELECT, nullptr, SoundEnvironment::Always);
-			SetInventoryMode(InventoryMode::None);
```
I believe this is a merge conflict. Did you merge `bugfixes_v20` branch or cherrypicked some commits from it to your PR?

TrainWrack replied:
Yes, bug fixes was merged. This will not be an issue once that branch gets merged.

TrainWrack committed:
Address comments

TrainWrack committed:
Address comments

TrainWrack committed:
Fix hide character for accessories and gunflash

Lwmte commented on TombEngine/Game/control/control.cpp:
```diff
@@ -310,6 +311,12 @@ GameStatus FreezePhase()
+	// Decrement gun flash counters for spectator mode
+	if (g_GameFlow->LastFreezeMode == FreezeMode::Spectator)
+	{
+		ReduceWeaponFlashesTimer(Lara);
```
Why do we need to decrement flash counter in freeze mode? Isn't it supposed to stay on? For example, when I put the game to freeze mode while Lara is shooting, I expect weapon gunflashes to stay visible.

TrainWrack replied:
They do, as long as its being called in Freeze mode continuously. 

Cause user can turn it on or off so I need a way to remove them after I stop calling them.

If not this, I would need to add a ClearGunFlash function.

Lwmte replied:
The point of original behaviour of gunflashes in freeze mode was that the state is maintained, i.e. all events triggered in normal game loop remain static in a freeze loop. Introducing flash counter decrement in a freeze mode breaks this concept.

TrainWrack replied:
I see what you mean. I can add a clear Gunflash function then.

TrainWrack replied:
Resolved

Lwmte commented on TombEngine/Game/Lara/lara_fire.cpp:
```diff
@@ -406,40 +430,72 @@ GameVector GetTargetPoint(ItemInfo& targetEntity)
 
-	case LaraWeaponType::Revolver:
-		return HolsterSlot::Revolver;
+void SpawnWeaponFlash(ItemInfo& laraItem, LaraWeaponType weaponType)
```
Can we reuse `SpawnWeaponFlash` for `ShootRightGun` and `ShootLeftGun` flipeffects?

TrainWrack replied:
The flip effects work independently with left or right side so we should not. It will needlessly complicate the function.

Lwmte replied:
Sorry for unresolving this, but I do not agree with you.
I believe that code which spawns a gunflash mesh should be generic between flipeffects and a script method. I looked at it and I see no problem in modifying it for that. The only culprit is dynamic light that you spawn, I believe you can also spawn dynamic light in lua instead of hardcoding it.

Lwmte replied:
Resolved myself

Lwmte commented on TombEngine/Game/Lara/lara_fire.cpp:
```diff
@@ -406,40 +430,72 @@ GameVector GetTargetPoint(ItemInfo& targetEntity)
+
+	if (isDualWield)
+	{
+		auto basePos = GetJointPosition(&laraItem, LM_RHAND).ToVector3();
```
To me it seems weird that dual wield weapons would spawn a dynamic light at the right hand position. Probably get both left and right hand joint positions and average between them, so that light appears in the middle?

TrainWrack replied:
addressed

TrainWrack committed:
Cleanup GunFlash handling

TrainWrack committed:
Correct UpdateGunFlash

Lwmte committed:
Merge branch 'develop' into pr/1664

Lwmte commented on TombEngine/Renderer/RendererDrawEffect.cpp:
```diff
@@ -1240,16 +1240,18 @@ namespace TEN::Renderer
-		if (Lara.Control.Weapon.GunType == LaraWeaponType::Flare)
+		auto flashWeaponType = Lara.RightArm.GunFlashType;
+
+		if (flashWeaponType == LaraWeaponType::None || flashWeaponType == LaraWeaponType::Flare)
```
You introduced a bug here. Because you compare only right arm gunflash, flipeffect that triggers gunflash for the left hand only, stopped working.

Lwmte replied:
Resolved myself

Lwmte committed:
Reuse SpawnWeaponFlash for legacy flipeffects

Lwmte committed:
Add relative mouse mode to gameplay loop

TrainWrack committed:
Expose WeaponGunFlash enums

TrainWrack committed:
Allow single gunflashes for single pistols

TrainWrack committed:
Safeguard SetSkin

TrainWrack committed:
6 probe collision, allow gunflash for default

6 probe collision check,
Allow gunflash for default weapon
Allow input of camera rotation on right when holding R2, Right Click

TrainWrack committed:
Camera cleanup, reset meshes on weapon default

Remove helper functions for existing Vec3 operands
Reset meshes for weapons when default is selected
reduce mouse senstivity default
Remove onExit from all menus

Lwmte commented on Scripts/Engine/PhotoMode/Accessories.lua:
```diff
@@ -0,0 +1,18 @@
+
+local Accessories =
+{
+    { name = "None",   objID = nil, meshIndices = {} },
```
Remove 2 extra spaces before `objID`.

TrainWrack replied:
Aligned

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+    local state   = States.Get()
+    local camRoom = state.cameraMesh:GetRoomNumber()
+    local front   = Camera.GetDirection():Normalize()
+    local right = front:Cross(WORLD_UP):Normalize()
```
Indentation is incorrect, either align all `=`s or remove alignment altogether.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+    local right = front:Cross(WORLD_UP):Normalize()
+    local up    = right:Cross(front):Normalize()
+
+    local dirs = {
```
Coding conventions violation. `{` should be on a new line.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+end
+
+function Camera.Detach()
+    pcall(TEN.View.ResetObjCamera)
```
Why `pcall` is used here?

TrainWrack replied:
Removed all pcalls

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+        local dx = newCam.x - origin.x
+        local dy = newCam.y - origin.y
+        local dz = newCam.z - origin.z
+        local distSq = dx * dx + dy * dy + dz * dz
```
There is a `Vec3::Distance` method, please use it instead recreating it here.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+end
+
+function Camera.MoveForward(speed)
+    local state = States.Get()
```
Indentation misalignment

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+    local camPos = state.cameraMesh:GetPosition()
+    local tgtPos = state.cameraTarget:GetPosition()
+    local offset = TEN.Vec3(tgtPos.x - camPos.x, tgtPos.y - camPos.y, tgtPos.z - camPos.z)
+    local rotated = offset:Rotate(TEN.Rotation(0, angle, 0))
```
Indentation misalignment

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+
+-- Maximum elevation angle (degrees from horizontal) the look-at vector is
+-- allowed to reach.  Staying below 90° prevents the camera from flipping.
+local PITCH_LIMIT = 88.0
```
Either put a constant in the beginning of the file or create separate Constants.lua for all constants.

TrainWrack replied:
done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+    local oz = tgtPos.z - camPos.z
+
+    -- Keep the camera-to-target distance constant.
+    local dist = math.sqrt(ox * ox + oy * oy + oz * oz)
```
Use `Vec3::Distance` method.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Camera.lua:
```diff
@@ -0,0 +1,291 @@
+    local dist = math.sqrt(ox * ox + oy * oy + oz * oz)
+    if dist < 0.001 then return end
+
+    local hDist = math.sqrt(ox * ox + oz * oz)
```
Use `Utils.CalculateHorizontalDistance`.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Configuration.lua:
```diff
@@ -0,0 +1,268 @@
+
+local Configuration = {}
+
+-- ============================================================================
```
Excessive formatting in comments, probably it's better to remove them?

TrainWrack replied:
Ring inventory also had them. I think we can keep them.

Lwmte commented on Scripts/Engine/PhotoMode/Configuration.lua:
```diff
@@ -0,0 +1,268 @@
+    meshIndex        = 0,
+    targetIndex      = 0,
+    defaultMoveSpeed = 50,
+    defaultLookSpeed  = 8.0,
```
Indent misalignment

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Configuration.lua:
```diff
@@ -0,0 +1,268 @@
+    minRadius      = 1,
+    maxRadius      = 20,
+    radiusStep     = 1,
+    defaultIntensity = 0.5,
```
Indent misalignment

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Configuration.lua:
```diff
@@ -0,0 +1,268 @@
+    sizeInactive  = TEN.Vec2(4, 4),                        -- size when not selected
+    colorActive   = Settings.ColorMap.neutral,
+    colorInactive = Settings.ColorMap.dimmed,
+    rotation  = 0,
```
Indent misalignment

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Configuration.lua:
```diff
@@ -0,0 +1,268 @@
+
+Configuration.HeaderSprites =
+{
+    objectID      = TEN.Objects.ObjID.PHOTOMODE_SPRITES, -- object that owns the sprites
```
Please format comments according to coding conventions (first capital letter, full stop at the end)

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Input.lua:
```diff
@@ -0,0 +1,286 @@
+        end
+    end
+    
+
```
Please remove extra line.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/InputHelpers.lua:
```diff
@@ -0,0 +1,65 @@
+local PULSE_DELAY = 0.25
+local FPS = 30
+-- Acceleration thresholds for option left/right navigation (hold to repeat faster).
+local ACCEL_INITIAL_DELAY = PULSE_DELAY   -- seconds after first press before repeating begins
```
Please put all constants into Constants.lua

TrainWrack replied:
These are only used in this file

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+
+local COLOR_MAP  = Configuration.ColorMap
+local SOUND_MAP  = Configuration.SoundMap
+local ALPHA_MAX  = 255
```
As above, put constants into Constants.lua.

TrainWrack replied:
These are used in this file only

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+
+    Menus[menuName] =
+    {
+        name           = menuName,
```
Please align indentation.

TrainWrack replied:
Constants.lua made sense for ring inventory as many constants were shared across files. These ones are only for this file so should stay here to avoid searching for constants while reviewing the code.

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+            menu.titleString, pos, menu.titleTextScale,
+            ColorCombine(menu.titleTextColor, actualTransparency),
+            menu.titleTranslate, menu.titleTextFlags
+        )
```
Please put bracket on the previous line.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+            ColorCombine(menu.titleTextColor, actualTransparency),
+            menu.titleTranslate, menu.titleTextFlags
+        )
+        TEN.Strings.ShowString(titleNode, 1 / 30)
```
1 / FPS?

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+            else
+                itemNode:SetFlags(menu.itemsTextFlags)
+            end
+            TEN.Strings.ShowString(itemNode, 1 / 30)
```
1 / FPS?

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+                menu.optionsTextScale,
+                ColorCombine(menu.optionsTextColor, actualTransparency),
+                menu.optionsTranslate
+            )
```
Please put bracket on the previous line.

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Menu.lua:
```diff
@@ -0,0 +1,816 @@
+            else
+                optNode:SetFlags(menu.optionsTextFlags)
+            end
+            TEN.Strings.ShowString(optNode, 1 / 30)
```
1 / FPS?

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Outfits.lua:
```diff
@@ -0,0 +1,68 @@
+        meshVisible = "all",
+    },
+
+        { name = "Classic TR2",
```
As above + indentation

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/Outfits.lua:
```diff
@@ -0,0 +1,68 @@
+    -- unlocked:          true/nil = outfit is visible in the menu.
+    --                    false    = hidden until PhotoMode.UnlockOutfit(name) is called.
+
+    { name = "Classic TR4",
```
Coding conventions violation. Opening curly bracket should be on a separate line.

TrainWrack replied:
These are temporary entries. Bur adjusted regardless.

Lwmte commented on Scripts/Engine/PhotoMode/Outfits.lua:
```diff
@@ -0,0 +1,68 @@
+        meshVisible = "all",
+    },
+
+    { name = "Remastered",
```
As above and further

TrainWrack replied:
These are temporary entries. But adjusted regardless.

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+
+
+-- Photo Module Start
+local Borders   = require("Engine.PhotoMode.SpriteBorders")
```
Indentation

TrainWrack replied:
Done

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+            show = ok and (count > 0 or Settings.Character.allWeapons)  -- Show if in inventory or if allWeapons setting is enabled.
+        end
+        if show then
+            local idx             = #names + 1
```
I believe indentation is wrong here

TrainWrack replied:
fioxed

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    local modePreset = cfg.modes[state.dofMode]
+    local mode = modePreset and modePreset.mode or TEN.View.DOFMode.NONE
+    if mode == TEN.View.DOFMode.NONE then
+        pcall(function() TEN.View.SetDOF(TEN.View.DOFMode.NONE) end)
```
Why `pcall` is used here and below?

TrainWrack replied:
Removed

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+            math.floor(128 + (c.r - 128) * i),
+            math.floor(128 + (c.g - 128) * i),
+            math.floor(128 + (c.b - 128) * i)
+        )
```
Bracket on the previous line

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- Restore skinned mesh to entry state.
+    if state.appliedSkinnedMesh then
+        if snap and snap.skinnedMeshObject then
+            pcall(function() Lara:SwapSkinnedMesh(snap.skinnedMeshObject, snap.skinnedMeshIndex) end)
```
pcall?

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- Adjust holster slots based on which visual slots the weapon occupies.
+    -- Clear the slots that are now visually shown as drawn; retain the rest.
+    -- For "none" (default), restore entry snapshot holster state.
+    pcall(function()
```
Can we offload this function from an in-place method to a separate body to de-clutter the code?

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    state.weaponIndex     = 1
+    state.expressionIndex = 1
+    state.accessoryIndex  = 1
+    state.gunflashEnabled   = false
```
Indentation

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+        state.lightPos        = TEN.Vec3(state.entryLight.pos.x, state.entryLight.pos.y, state.entryLight.pos.z)
+        state.lightRadius     = state.entryLight.radius
+        state.lightShadows    = state.entryLight.shadows
+        state.lightIntensity   = state.entryLight.intensity
```
Indentation

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+        if     name == "pm_enabled" then state.lightEnabled    = IndexToBool(m:GetCurrentOptionIndex())
+        elseif name == "pm_source"  then state.lightSource     = m:GetCurrentOptionIndex()
+        elseif name == "pm_radius"  then state.lightRadius     = OptionIndexToValue(m:GetCurrentOptionIndex(), cfg.Light.minRadius, cfg.Light.radiusStep)
+        elseif name == "pm_intensity" then state.lightIntensity = OptionIndexToValue(m:GetCurrentOptionIndex(), cfg.Light.minIntensity, cfg.Light.intensityStep)
```
Indentation

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    local outfitNames = BuildFilteredOutfitNames()
+    local weaponNames = BuildFilteredWeaponNames()
+    local characterItems = {
```
Curly bracket on a new line

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    -- EFFECTS menu (Lens + Depth of Field)
+    -- ================================================================
+    local effectsItems = {
```
As above

TrainWrack committed:
Unhide UI with Escape

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+          currentOption = ValueToOptionIndex(state.fov, cfg.Lens.minFOV, cfg.Lens.fovStep)},
+        { itemName = "pm_roll", options = NumberRange(cfg.Lens.minRoll, cfg.Lens.maxRoll, cfg.Lens.rollStep),
+          currentOption = ValueToOptionIndex(state.roll, cfg.Lens.minRoll, cfg.Lens.rollStep)},
+    }
```
Add new line after closing bracket

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    -- FILTERS menu (Post-process + Frame)
+    -- ================================================================
+    CreateMenu(MENU_FILTERS, {
```
Curly bracket on a new line

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    -- LIGHT menu
+    -- ================================================================
+    CreateMenu(MENU_LIGHT, {
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    -- UI menu
+    -- ================================================================
+    CreateMenu(MENU_UI, {
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    -- ================================================================
+    -- Set up headers (STEP_LEFT / STEP_RIGHT to navigate)
+    -- ================================================================
+    Menu.SetHeaders({
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+
+    -- Store entry light state
+    local camPos = state.cameraMesh:GetPosition()
+    state.entryLight = {
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    local COLOR_SWATCH_SPRITE = 8
+
+    -- Descriptor table for each colour-picking item
+    local entries = {
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/States.lua:
```diff
@@ -0,0 +1,319 @@
+-- Runtime State (reset on entry, restored on exit)
+-- ============================================================================
+
+local State = {
```
As above

Lwmte commented on Scripts/Engine/PhotoMode/PhotoMode.lua:
```diff
@@ -0,0 +1,1843 @@
+    States.Get().timeInPhotoMode = 0
+
+    TEN.Sound.PlaySound(Configuration.SoundMap.menuClose)
+    TEN.Input.ClearAllKeys()
```
This causes input collision on exit, because ESC key is treated as clicked, and inventory is immediately opened.

TrainWrack replied:
Removed

Lwmte committed:
Fix random enemy gunflash rotation in freeze mode and introduce gunflash rotation for Lara gunflashes

TrainWrack committed:
Address comments, User RAY for collision

Address formatting,
fix gunflash spawn
Use Ray for collision

TrainWrack committed:
Fix title lock, Mouse scroll added for menus

Lwmte committed:
Added player collision

Lwmte committed:
Add camera sliding on collision

Lwmte committed:
Make SaveScreenshot return path string

Lwmte committed:
Add screenshot functionality

Lwmte committed:
Formatting, movement optimizations

Lwmte committed:
Fix text message

Lwmte committed:
Formatting, change look key to take screenshot

Lwmte committed:
Fix relative mouse mode in fullscreen

Lwmte committed:
Update Strings.lua

Lwmte committed:
Adjust hint timeout

Lwmte committed:
Update doc formatting

Lwmte committed:
Align function declarations

Lwmte committed:
Trigger look to reset thumbstick camera

Lwmte committed:
Update PhotoMode.lua

Lwmte committed:
Decopypaste input code

Lwmte committed:
Align menu values to the right

Lwmte committed:
Fix swatch position

TrainWrack commented on TombEngine/Game/control/control.cpp:
```diff
@@ -322,6 +323,7 @@ GameStatus FreezePhase()
 
 		UpdateAllItems();
 		UpdateGlobalLensFlare();
+		Weather.Update();
```
@Lwmte Should it be Weather.Update(true) here?

Lwmte replied:
Yeah, sorry, my mistake. Fixed.

TrainWrack committed:
Simplify vector math, fix borders drawing in screenshots

Fix global functions in ring inventory
Add screenshot sound
Simplify vector math
Fix borders in screenshots

Lwmte committed:
Fixed weather update in freeze mode

Lwmte committed:
Merge branch 'develop' into pr/1664

Lwmte committed:
Make "Hide UI" a state option, not a toggle option

Lwmte committed:
Rearrange menus, correct hints, formatting

Lwmte committed:
Highlight with header color

TrainWrack committed:
Add PhotoModeData file

TrainWrack committed:
Address comments

Add to AutoExec
Update Changelog
Remove magic numbers
Update PhotoModeData.lua
Update docs for PhotoMode

TrainWrack committed:
Adjust changelog

Lwmte committed:
Rewrite menu alignment

Lwmte committed:
Fix menu title position

Lwmte committed:
Wrap menus and non-numerical options

TrainWrack committed:
Adjust color swatch size

TrainWrack committed:
Fix frame scaling

Lwmte committed:
Move documentation out of the main module

Lwmte committed:
Reformat docs

TrainWrack committed:
Adjust docs

TrainWrack committed:
Add sections of PhotoModeSetup

Lwmte commented on TombEngine/Game/control/los.cpp:
```diff
@@ -160,11 +160,14 @@ bool GetTargetOnLOS(GameVector* origin, GameVector* target)
+		{
 			SoundEffect(SFX_TR4_REVOLVER_FIRE, nullptr);
-
+			Lara.RightArm.GunFlashType = LaraWeaponType::Revolver;
```
Why this was added? GetTargetOnLOS is used for lasersight mode which doesn't involve any gunflashes.

TrainWrack replied:
Cause we are adding timer to GunFlash on line 160. 

`Lara.RightArm.GunFlash = Weapons[(int)Lara.Control.Weapon.GunType].FlashTime;`

The reason GunFlashType was added cause anywhere GunFlash was populated, the GunFlashType was added

Lwmte commented on TombEngine/Game/control/los.cpp:
```diff
@@ -160,11 +160,14 @@ bool GetTargetOnLOS(GameVector* origin, GameVector* target)
 		if (Lara.Control.Weapon.GunType == LaraWeaponType::Crossbow)
 		{
 			FireCrossBowFromLaserSight(*LaraItem, origin, &target2);
+			Lara.RightArm.GunFlashType = LaraWeaponType::Crossbow;
```
As above.

TrainWrack replied:
As above

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -663,6 +665,90 @@ std::unique_ptr<Moveable> LaraObject::GetPlayerInteractedMoveable() const
+// @treturn Objects.WeaponType Back holster weapon type.
+// @usage
+// local left, right, back = Lara:GetHolsterWeapon()
+std::tuple<LaraWeaponType, LaraWeaponType, LaraWeaponType> LaraObject::GetHolsterWeapon() const
```
This function must be renamed to `GetHolsterWeaponTypes` because we are returning 3 different weapon types.

TrainWrack replied:
Addressed

For lua kept it simple to SetHolsterWeapons /  GetHolsterWeapons

Lwmte replied:
Not agreed. We already have functions `GetWeaponType` and `SetWeaponType` for LaraObject. These new functions should correspond to that to keep Lua API consistent.

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -663,6 +665,90 @@ std::unique_ptr<Moveable> LaraObject::GetPlayerInteractedMoveable() const
+// @usage
+// Lara:SetHolsterWeapon(WeaponType.PISTOLS, WeaponType.PISTOLS, WeaponType.SHOTGUN)
+// Lara:SetHolsterWeapon(nil, nil, WeaponType.NONE) -- Clear back holster only.
+void LaraObject::SetHolsterWeapon(TypeOrNil<LaraWeaponType> left, TypeOrNil<LaraWeaponType> right, TypeOrNil<LaraWeaponType> back)
```
As above, but `SetHolsterWeaponTypes`.

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -663,6 +665,90 @@ std::unique_ptr<Moveable> LaraObject::GetPlayerInteractedMoveable() const
+// Lara:ResetHair()
+void LaraObject::ResetHair()
+{
+	using namespace TEN::Effects::Hair;
```
Such local namespace definitions is not okay. Either move this using to the top of the file, or use full namespace in place while calling the method.

TrainWrack replied:
Resolved

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -663,6 +665,90 @@ std::unique_ptr<Moveable> LaraObject::GetPlayerInteractedMoveable() const
+	SpawnWeaponFlash(*_moveable, convertedFlag, weaponType);
+}
+
+/// Clear the muzzle flash.
```
Possibly it's better to document this function as "Clear all currently active muzzle flashes", because it clears both of them.

TrainWrack replied:
Resolved

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Lara/LaraObject.cpp:
```diff
@@ -763,21 +849,23 @@ sol::table LaraObject::GetSkin(sol::this_state s)
 		if (!isValidObjectID(skin.value()))
 		{
-			TENLog("SetSkin: skin object ID " + std::to_string(skin.value()) + " is invalid or not loaded.", LogLevel::Warning, LogConfig::All);
+			TENLog("SetSkin: skin object ID " + GetObjectName(skin.value()) + " (" + std::to_string(skin.value()) + ")" + " is invalid or not loaded.", LogLevel::Warning, LogConfig::All);
```
There's a lot of boilerplate warnings down the line. I suggest to use a single subfunction to construct a warning message in all places.

TrainWrack replied:
Resolved

Lwmte commented on TombEngine/Scripting/Internal/TEN/Objects/Moveable/MoveableObject.cpp:
```diff
@@ -1043,15 +1045,31 @@ void Moveable::ShatterMesh(int meshId)
 // @treturn bool Mesh swap status.
-bool Moveable::GetMeshSwapped(int meshId) const
+// @treturn[opt] Objects.ObjID Object slot ID the mesh was swapped from. Nil if not swapped.
+std::tuple<bool, sol::optional<GAME_OBJECT_ID>> Moveable::GetMeshSwapped(int meshId) const
```
Since function declaration changed, did you test old-styled calls (without the new return value) to this function to make sure they are compatible?

TrainWrack replied:
Yes, if the user is just calling the function with one variable

local test = Moveable:GetMeshSwapped(2); it will only return false/true. Only when the function is called with two variables

local test, test2 = Moveable:GetMeshSwapped(2); will it return the object ID.

Lwmte committed:
Solve thumbstick camera issues in native code instead of script

TrainWrack committed:
Address comments

TrainWrack committed:
Change HolsterTypes

Lwmte committed:
Fix indentation

Lwmte committed:
Rename ClearGunFlash to ClearGunFlashes

Lwmte committed:
Update CHANGELOG.md

Lwmte committed:
Update CHANGELOG.md
